### ISSUE TITLE
destruct vs refine (match ... end)

### ISSUE THREAD

**Issue by anton-trunov** [coq/coq#9203]
#### Version

8.8.2

#### Operating system

macOS 10.14

#### Description of the problem

```coq
Inductive foo : bool -> Type := foot : foo true.

Goal forall big : foo true, big = foot.
  intros big.
  Fail destruct big.
  refine (match big with foot => _ end).
  reflexivity.
Qed.
```

Should not `destruct x` be at least as powerful as
```coq
refine (match x | C1 ... => _ | ... | Cn ... => _ end)
```
?

Thank you.

**Comment by gares** [comment]
I think they are incomparable. Destruct may generalize context items. Refine may prove impossible cases (not sure destruct does that).

**Comment by anton-trunov** [comment]
Yes, I understand, but I guess I was trying to say that it feels a bit like a dissonance when I can do the proof explicitly (with the great help of the elaborator)
```coq
Definition bar : forall big : foo true, big = foot :=
  fun big => match big with
             | foot => eq_refl
             end.
```
and I can't write it using the (seemingly) corresponding tactics
```coq
Lemma bar : forall big : foo true, big = foot.
  intros big.
  Fail destruct big.
Abort.
```

**Comment by gares** [comment]
I agree, I think the tactic should be strictly more powerful, and be based on top of the elaborator.

**Comment by ppedrot** [comment]
I agree as well, but expect backward compatibility issues.

**Comment by anton-trunov** [comment]
@ppedrot Do you happen to have an example of what might break? I'd be very much interested.

>  I think the tactic should be strictly more powerful, and be based on top of the elaborator.

It would be great if `intros []` was also based on this (see e.g. #7972).
